samza-core/src/main/java/org/apache/samza/operators/triggers/RepeatingTriggerImpl.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.samza.operators.triggers;
+
+import org.apache.samza.operators.data.MessageEnvelope;
+
+/**
+ * Implementation class for a {@link RepeatingTrigger}
+ */
+public class RepeatingTriggerImpl<M extends MessageEnvelope> implements TriggerImpl<M> {
+
+  private final Trigger<M> underlyingTrigger;
+
+  private TriggerImpl<M> underlyingTriggerImpl;
+
+  public RepeatingTriggerImpl(RepeatingTrigger<M> repeatingTrigger) {
+    this.underlyingTrigger = repeatingTrigger.getTrigger();
+    this.underlyingTriggerImpl = TriggerImpls.createTriggerImpl(underlyingTrigger);
+  }
+
+  private TriggerCallbackHandler createNewHandler(TriggerCallbackHandler handler) {
+    return new TriggerCallbackHandler() {
+      @Override
+      public void onTrigger() {
+          //re-schedule the underlying trigger for execution again.
+          underlyingTriggerImpl = TriggerImpls.createTriggerImpl(underlyingTrigger);
```

Comment:
- We construct a wrapped handler object that proxies the invocation to the parent (after cancelation).
- Consider a case `Repeating(Triggers.count(5))`. 

In this case, when the handler for the `CountTrigger` fires, the `RepeatingTrigger` will `cancel` the `CountTrigger`, create a new instance of the `CountTriggerImpl` for a potential future firing, and propagate the firing by calling `onTrigger` on it's own handler (passed to it upstream).